### PR TITLE
Updated type of action parameter for DataSnapshot#forEach

### LINKED ISSUE #6368: rtdb(types): DataSnapshot#forEach - action parameter has null key
### [REQUIRED] Describe your environment

  * Operating System version: N/A
  * Browser version: N/A
  * Firebase SDK version: 9.8
  * Firebase Product: database

### [REQUIRED] Describe the problem

When using `DataSnapshot#forEach` the parameter available to the action closure is a `DataSnapshot` which is correct, but the `key` property of this value has the type `string | null`, but as its a child of the iterating snapshot it can't be the root node so its key can never be `null`.

#### Steps to reproduce:

```js
const node = await getDatabase().ref().get()
node.forEach(child => {
   const key = child.key
   // key is possibly null so have to cast etc.
})
```

#### Relevant Code:

https://github.com/firebase/firebase-js-sdk/blob/efe2000fc499e2c85c4e5e0fef6741ff3bad2eb0/packages/database-types/index.d.ts#L25

https://github.com/firebase/firebase-js-sdk/blob/efe2000fc499e2c85c4e5e0fef6741ff3bad2eb0/packages/firebase/compat/index.d.ts#L5821-L5823

https://github.com/firebase/firebase-js-sdk/blob/efe2000fc499e2c85c4e5e0fef6741ff3bad2eb0/packages/database-compat/src/api/Reference.ts#L167

#### Fix

Happy to do a PR for this - I just remember it being a bit complicated [last time](https://github.com/firebase/firebase-js-sdk/issues/6071#issuecomment-1074261971) I did this. Is the info in that issue still correct - the function signature in the above 3 places just needs to be changed?

Would changing the function signature to `forEach(action: (a: DataSnapshot & { key: string }) => boolean | void): boolean;` be OK for a fix?

### PR DESCRIPTION
**Pull request by rhodgkins**
Fixes #6368

### CODE REVIEW

File: .changeset/silent-carrots-bathe.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+'@firebase/database': patch
+'@firebase/database-compat': patch
+'@firebase/database-types': patch
+'firebase': patch
```
[the remark this answers is not in the document]
Comment (rhodgkins) [reply .changeset/silent-carrots-bathe.md:5]: Ah I only added it as [last time I did a type change](https://github.com/firebase/firebase-js-sdk/pull/6090#pullrequestreview-920666330) I got asked to add one.

[the remark this answers is not in the document]
Comment (maneesht) [reply .changeset/silent-carrots-bathe.md:5]: Yes, a changeset should be added, I just meant line 5 should be removed

**Review by maneesht — APPROVED** [review]

**Comment by rhodgkins** [comment]
Thanks @maneesht - the failing API report check, do I need to do anything?

**Comment by maneesht** [comment]
@rhodgkins - not sure if it will help, but your forked repo is behind, can you update it?

**Comment by rhodgkins** [comment]
@maneesht rebased but no change. According to [the logs](https://github.com/firebase/firebase-js-sdk/runs/7151650485?check_suite_focus=true#step:2:13) its missing a token...?

**Comment by maneesht** [comment]
Looks like there's an issue related to github actions. We have a fix incoming.

**Review by jsdt — APPROVED** [review]

**Comment by rhodgkins** [comment]
Thanks @jsdt - @maneesht I rebased to include your fix (#6409), but still failing unfortunately...

**Comment by rhodgkins** [comment]
@maneesht what else do I need to do to get this merged...? Rather not have this sit here for months as its such a small change!

**Comment by maneesht** [comment]
@rhodgkins - can you please run `yarn build` locally, so that it generates the correct md file, and then commit and push the results?

**Review by hsubox76 — APPROVED** [review]

**Pull request merged**